src/Nri/Ui/FocusTrap/V1.elm

```diff
@@ -20,34 +19,68 @@ import Json.Decode as Decode exposing (Decoder)
 import Task
 
 
-{-| -}
-only : { focusSelf : msg } -> List (Html.Attribute msg)
-only { focusSelf } =
-    [ onKeyDownPreventDefault
-        [ Key.tab focusSelf
-        , Key.tabBack focusSelf
-        ]
-    , class "focus-trap__only-focusable-element"
-    ]
+{-| Defines how focus will wrap in reponse to tab keypresses in a part of the UI.
 
+The ids referenced here are expected to correspond to elements in the container
+we are applying the focus trap on.
+
+-}
+type alias FocusTrap msg =
+    { firstId : String
+    , lastId : String
+    , focus : String -> msg
+    }
+
+
+{-| Attach this attribute to add a focus trap to an HTML element.
+-}
+toAttribute : FocusTrap msg -> Html.Attribute msg
+toAttribute { firstId, lastId, focus } =
+    onTab <|
+        \elementId shiftKey ->
+            -- if the user tabs back while on the first id,
+            -- we want to wrap around to the last id.
+            if elementId == firstId && shiftKey then
+                Decode.succeed
+                    { message = focus lastId
+                    , preventDefault = True
+                    , stopPropagation = False
+                    }
+
+            else if elementId == lastId && not shiftKey then
+                -- if the user tabs forward while on the last id,
+                -- we want to wrap around to the first id.
+                Decode.succeed
+                    { message = focus firstId
+
+                    -- Will this preventDefault break anything
+                    -- if the element is an input
+                    -- or dropdown (if the element has key-based
+                    -- behavior already)?
```

Comment:
I'm pretty sure it won't break anything because:
  - preventDefault should only affect whether or not the native behavior runs, but it shouldn't mess up with event propagation
  - we are handling the event as it is bubbling up, and at this point event handlers on descendant elements should have already been fired

With that said, testing locally before merging shouldn't hurt. I'll try it out!

Reply:
I just added a text input to the example in the style guide and verified that we can still attach event handlers to it as usual. All seems to be working fine!